Add bulk upgrade CLI

hu55a1n1 opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Closes: #763 

## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->
Add a bulk upgrade CLI command that takes a source chain's id as input and upgrades clients for all chains in config target the specified source chain client.
______

For contributor use:

- [ ] Updated the __Unreleased__ section of [CHANGELOG.md](https://github.com/informalsystems/ibc-rs/blob/master/CHANGELOG.md) with the issue.
- [ ] If applicable: Unit tests written, added test to CI.
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

hu55a1n1 commented on relayer-cli/src/conclude.rs:
```diff
@@ -69,11 +69,7 @@ pub fn exit_with(out: Output) {
-    } else {
-        println!("{}: {}", out.status, out.result);
-    }
+    println!("{}", String::from(out));
```
Just a minor suggestion - I think it would be nice to write success messages to `stdout` (using `println!()`) and errors to `stderr` (using `eprintln!()`).

hu55a1n1 commented on relayer-cli/src/conclude.rs:
```diff
@@ -207,6 +203,24 @@ impl Output {
         Output::with_success().with_msg(msg)
     }
 
+    pub fn combined(outputs: Vec<Self>) -> Output {
```
I am not sure if this is the best abstraction for dealing with multiple outputs, since the way this function is implemented, even if a single `Output` has a status `Error` the combined output is created with a status `Error`. 
Another solution would be to create a `fn print(&self)` function that would print an output (to stdout/stderr) and the caller gets to decide how/when to call it.

romac replied:
Yeah the `Output` type was initially not meant for displaying the output incrementally, but rather to provide the final output of the command on stdout for consumption by scripts etc. when JSON output is enabled. We currently do incremental output with the `tracing` crate which outputs to stderr. While this is not perfect, I think we should stick with this model for the purpose of this PR and try to revisit this in the future. What do you think?

romac replied:
To make it clearer, my suggestion here is to use `tracing` to display status updates incrementally (eg. on each client update), collect the results somewhere in a custom datatype with a manual `Display` instance, and use `Output` on that.

hu55a1n1 replied:
It makes a lot of sense to think of the `Output` as the final output of a command. 👍 I have modified the code to remove `Output::combined()` and instead accumulate the results in a newtype which I use to create a final `Output`. 
Here's what the final output looks like -> 
```bash
# two outer errors
.
├─ config error: missing chain for id (07-tendermint-0) in configuration file
└─ config error: missing chain for id (07-tendermint-1) in configuration file

# one inner errors
.
└─.
  └─ chain runtime error: failed while trying to upgrade client id 07-tendermint-0 with error: failed while querying src chain

# one outer for host-0 and two inners errors both for host-1
.
├─ config error: missing chain for id (07-tendermint-0) in configuration file
└─.
  ├─ chain runtime error: failed while trying to upgrade client id 07-tendermint-1 with error: failed while querying src chain
  └─ chain runtime error: failed while trying to upgrade client id 07-tendermint-1 with error: failed while querying src chain

# one outer and 2 inner errors each from different hosts
.
├─ config error: missing chain for id (07-tendermint-0) in configuration file
├─.
│ └─ chain runtime error: failed while trying to upgrade client id 07-tendermint-0 with error: failed while querying src chain
└─.
  └─ chain runtime error: failed while trying to upgrade client id 07-tendermint-0 with error: failed while querying src chain

# one outer error, one inner error and one success message (Vec<IbcEvent> serialized) each from different hosts 
.
├─ config error: missing chain for id (07-tendermint-0) in configuration file
├─.
│ └─ chain runtime error: failed while trying to upgrade client id 07-tendermint-0 with error: failed while querying src chain
└─.
  └─ [
    UpgradeClient(
        UpgradeClient(
            Attributes {
                height: Height {
                    revision: 0,
                    height: 0,
                },
                client_id: ClientId(
                    "07-tendermint-0",
                ),
                client_type: Tendermint,
                consensus_height: Height {
                    revision: 0,
                    height: 0,
                },
            },
        ),
    ),
]

# success message
[[{"UpgradeClient":{"height":{"revision_number":0,"revision_height":0},"client_id":"07-tendermint-0","client_type":"Tendermint","consensus_height":{"revision_number":0,"revision_height":0}}}],[{"UpgradeClient":{"height":{"revision_number":0,"revision_height":0},"client_id":"07-tendermint-0","client_type":"Tendermint","consensus_height":{"revision_number":0,"revision_height":0}}},{"UpgradeClient":{"height":{"revision_number":0,"revision_height":0},"client_id":"07-tendermint-0","client_type":"Tendermint","consensus_height":{"revision_number":0,"revision_height":0}}},{"UpgradeClient":{"height":{"revision_number":0,"revision_height":0},"client_id":"07-tendermint-0","client_type":"Tendermint","consensus_height":{"revision_number":0,"revision_height":0}}},{"UpgradeClient":{"height":{"revision_number":0,"revision_height":0},"client_id":"07-tendermint-0","client_type":"Tendermint","consensus_height":{"revision_number":0,"revision_height":0}}}]]
```
And here's the test used to generate this output -> 
```rust
#[cfg(test)]
mod tests {
    use super::*;
    use ibc::ics02_client::client_type::ClientType;
    use ibc::ics02_client::events::Attributes;
    use ibc_relayer::foreign_client::ForeignClientError;

    fn print_output_buffer(ob: OutputBuffer) {
        match ob.into_result() {
            Ok(events) => println!("{}", serde_json::to_string(&events).unwrap()),
            Err(e) => println!("{}", e),
        }
    }

    #[test]
    fn test_output_buffer() {
        print_output_buffer(OutputBuffer(vec![
            Err(Kind::Config
                .context(format!(
                    "missing chain for id ({}) in configuration file",
                    ClientId::new(ClientType::Tendermint, 0).unwrap()
                ))
                .into()),
            Err(Kind::Config
                .context(format!(
                    "missing chain for id ({}) in configuration file",
                    ClientId::new(ClientType::Tendermint, 1).unwrap()
                ))
                .into()),
        ]));

        print_output_buffer(OutputBuffer(vec![Ok(vec![Err(Kind::Runtime
            .context(ForeignClientError::ClientUpgrade(
                ClientId::new(ClientType::Tendermint, 0).unwrap(),
                "failed while querying src chain".to_owned(),
            ))
            .into())])]));

        print_output_buffer(OutputBuffer(vec![
            Err(Kind::Config
                .context(format!(
                    "missing chain for id ({}) in configuration file",
                    ClientId::new(ClientType::Tendermint, 0).unwrap()
                ))
                .into()),
            Ok(vec![
                Err(Kind::Runtime
                    .context(ForeignClientError::ClientUpgrade(
                        ClientId::new(ClientType::Tendermint, 1).unwrap(),
                        "failed while querying src chain".to_owned(),
                    ))
                    .into()),
                Err(Kind::Runtime
                    .context(ForeignClientError::ClientUpgrade(
                        ClientId::new(ClientType::Tendermint, 1).unwrap(),
                        "failed while querying src chain".to_owned(),
                    ))
                    .into()),
            ]),
        ]));

        print_output_buffer(OutputBuffer(vec![
            Err(Kind::Config
                .context(format!(
                    "missing chain for id ({}) in configuration file",
                    ClientId::new(ClientType::Tendermint, 0).unwrap()
                ))
                .into()),
            Ok(vec![Err(Kind::Runtime
                .context(ForeignClientError::ClientUpgrade(
                    ClientId::new(ClientType::Tendermint, 0).unwrap(),
                    "failed while querying src chain".to_owned(),
                ))
                .into())]),
            Ok(vec![Err(Kind::Runtime
                .context(ForeignClientError::ClientUpgrade(
                    ClientId::new(ClientType::Tendermint, 0).unwrap(),
                    "failed while querying src chain".to_owned(),
                ))
                .into())]),
        ]));

        print_output_buffer(OutputBuffer(vec![
            Err(Kind::Config
                .context(format!(
                    "missing chain for id ({}) in configuration file",
                    ClientId::new(ClientType::Tendermint, 0).unwrap()
                ))
                .into()),
            Ok(vec![Err(Kind::Runtime
                .context(ForeignClientError::ClientUpgrade(
                    ClientId::new(ClientType::Tendermint, 0).unwrap(),
                    "failed while querying src chain".to_owned(),
                ))
                .into())]),
            Ok(vec![Ok(vec![IbcEvent::UpgradeClient(
                Attributes::default().into(),
            )])]),
        ]));

        print_output_buffer(OutputBuffer(vec![
            Ok(vec![Ok(vec![IbcEvent::UpgradeClient(
                Attributes::default().into(),
            )])]),
            Ok(vec![Ok(vec![
                IbcEvent::UpgradeClient(Attributes::default().into()),
                IbcEvent::UpgradeClient(Attributes::default().into()),
                IbcEvent::UpgradeClient(Attributes::default().into()),
                IbcEvent::UpgradeClient(Attributes::default().into()),
            ])]),
        ]));
    }
}
```

romac replied:
Looks great! 💯

adizere commented on relayer-cli/src/commands/tx/client.rs:
```diff
@@ -163,3 +164,64 @@ impl Runnable for TxUpgradeClientCmd {
+        config: &config::Reader<CliApp>,
+        dst_chain_id: &ChainId,
+    ) -> Result<Vec<IbcEvent>, Error> {
+        let chains = ChainHandlePair::spawn(&config, &self.src_chain_id, dst_chain_id)?;
```
Note that here we are re-`spawn`-ing the chain runtime for the source chain every time the method `upgrade_chain` is called. Instead of using `ChainHandlePair::spawn` there may be slightly more efficient ways to do it. One alternative, for instance:

- in the caller (`TxUpgradeClientsCmd::run`) we spawn the chain runtime for the source chain a single time, using something like [`spawn_chain_runtime`](https://github.com/informalsystems/ibc-rs/blob/7712422a47c25563e8c211fcbf79fcaabec142ec/relayer-cli/src/cli_utils.rs#L38) which would return the chain handle for the source chain id.
- then we pass this chain handle (or a clone thereof) when calling `self.upgrade_client(&config, &chain_id, &src_chain_handle)`.

adizere commented on relayer-cli/src/commands/tx/client.rs:
```diff
@@ -163,3 +164,64 @@ impl Runnable for TxUpgradeClientCmd {
+            })
+            .collect();
+
+        debug_assert!(client_ids.len() < 2);
```
I'm curious what is this assert for?

hu55a1n1 replied:
I wrongly presumed there could be only one client per host-target pair.  😓 Fixed!

hu55a1n1 commented on relayer-cli/src/commands/tx/client.rs:
```diff
@@ -163,3 +164,64 @@ impl Runnable for TxUpgradeClientCmd {
+        debug_assert!(client_ids.len() < 2);
+
+        let mut events = vec![];
+        if let Some(client_id) = client_ids.first() {
```
Done 👍

adizere commented on relayer-cli/src/conclude.rs:
```diff
@@ -207,6 +203,26 @@ impl Output {
 
+    /// Combine multiple `Output`s into a single `Output` by concatenating their results. The
+    /// combined `Output` has status `Error` if even one of the constituent `Output`s is an error.
+    pub fn combined(outputs: impl IntoIterator<Item = Output>) -> Output {
```
I think the next step could be to try and test this PR. We can reuse the instruction from https://hermes.informal.systems/commands/upgrade/test.html but it would be good to ensure that:

1. the upgraded chain has multiple clients, which could be hosted on two other chains
2. one of the clients should be frozen, or we should have some condition to simulate the failure of upgrading at least a client, so we can make sure that the output we have here is appropriate.


I'm adding a ["needs guide update" label](https://github.com/informalsystems/ibc-rs/pulls?q=is%3Aopen+is%3Apr+label%3A%22needs+guide+update%22) to this PR to also flag that we might want to update the guide.

romac commented on relayer-cli/src/conclude.rs:
```diff
@@ -252,6 +248,16 @@ fn serialize_result(res: impl Serialize + std::fmt::Debug) -> serde_json::Value
     }
 }
 
+impl From<Output> for String {
```
Unless there is a good reason for extracting this code out of `exit_with`, can we move it back there?
I am a bit wary of having a `From` instance whose behavior depends on a runtime condition (ie. whether JSON output is enabled or not). Maybe we can instead move this code into a method on `Output`?

hu55a1n1 replied:
Done 👍 ...

hu55a1n1 commented:
I tried testing this using [hermes guide - testing client upgrade](https://hermes.informal.systems/commands/upgrade/test.html). I modified the procedure slightly to test with multiple host chains (i.e. ibc-1 & ibc-2) targeting a single chain (i.e. ibc-0) using the following commands -> 
```bash
$ cat ~/.hermes/config.toml
[global]
strategy = 'packets'
log_level = 'info'

[[chains]]
id = 'ibc-0'
rpc_addr = 'http://127.0.0.1:26657'
grpc_addr = 'http://127.0.0.1:9090'
websocket_addr = 'ws://localhost:26657/websocket'
rpc_timeout = '10s'
account_prefix = 'cosmos'
key_name = 'testkey'
store_prefix = 'ibc'
gas = 200000
fee_denom = 'stake'
fee_amount = 10
clock_drift = '5s'
trusting_period = '14days'

[chains.trust_threshold]
numerator = '1'
denominator = '3'

[[chains]]
id = 'ibc-1'
rpc_addr = 'http://127.0.0.1:26557'
grpc_addr = 'http://127.0.0.1:9091'
websocket_addr = 'ws://localhost:26557/websocket'
rpc_timeout = '10s'
account_prefix = 'cosmos'
key_name = 'testkey'
store_prefix = 'ibc'
gas = 200000
fee_denom = 'stake'
fee_amount = 10
clock_drift = '5s'
trusting_period = '14days'

[chains.trust_threshold]
numerator = '1'
denominator = '3'

[[chains]]
id = 'ibc-2'
rpc_addr = 'http://127.0.0.1:26457'
grpc_addr = 'http://127.0.0.1:9092'
websocket_addr = 'ws://localhost:26457/websocket'
rpc_timeout = '10s'
account_prefix = 'cosmos'
key_name = 'testkey'
store_prefix = 'ibc'
gas = 200000
fee_denom = 'stake'
fee_amount = 10
clock_drift = '5s'
trusting_period = '14days'

[chains.trust_threshold]
numerator = '1'
denominator = '3'
$ ./scripts/dev-env ~/.hermes/config.toml ibc-0 ibc-1 ibc-2
$ ./scripts/dev-env ~/.hermes/config.toml ibc-0 ibc-1 ibc-2
WARNING: ./scripts/setup-chains will DELETE the '/home/hussaini/Documents/ibc-rs/data' folder.
> Do you wish to continue? (y/n): y
GAIA VERSION INFO: v4.2.1
Generating gaia configurations...
Creating gaiad instance: home=./data | chain-id=ibc-0 | p2p=:26656 | rpc=:26657 | profiling=:6060 | grpc=:9090 | samoleans=:100000000000
Change settings in config.toml file...
Start gaia on grpc port: 9090...
Balances for validator 'cosmos1hp2dcwqycyfcpn9xtg4tn9hv65klfks33szfgv' @ 'tcp://localhost:26657'
balances:
- amount: "0"
  denom: stake
pagination:
  next_key: null
  total: "0"
Balances for user 'cosmos1x9ueqshn8grl7qtj8agp6j359f4snm8yfgznd9' @ 'tcp://localhost:26657'
balances:
- amount: "100000000000"
  denom: samoleans
- amount: "100000000000"
  denom: stake
pagination:
  next_key: null
  total: "0"
Balances for user 'cosmos1qhv4lcqm772h4d9k2dfrsx2gvyfeh0gr8jf4zv' @ 'tcp://localhost:26657'
balances:
- amount: "100000000000"
  denom: samoleans
- amount: "100000000000"
  denom: stake
pagination:
  next_key: null
  total: "0"
Creating gaiad instance: home=./data | chain-id=ibc-1 | p2p=:26556 | rpc=:26557 | profiling=:6061 | grpc=:9091 | samoleans=:100000000000
Change settings in config.toml file...
Start gaia on grpc port: 9091...
Balances for validator 'cosmos196rtz4g28e5mc3nuwtd0j4tw5uxkkye6mv5hs7' @ 'tcp://localhost:26557'
balances: []
pagination:
  next_key: null
  total: "0"
Balances for user 'cosmos1ph9cnw62mlgd997whfjdrgcuaq3a7usypt090f' @ 'tcp://localhost:26557'
balances: []
pagination:
  next_key: null
  total: "0"
Balances for user 'cosmos1t3mmfnrprxl46gn4cru654gktzu5t7uf64flst' @ 'tcp://localhost:26557'
balances: []
pagination:
  next_key: null
  total: "0"
Creating gaiad instance: home=./data | chain-id=ibc-2 | p2p=:26456 | rpc=:26457 | profiling=:6062 | grpc=:9092 | samoleans=:100000000000
Change settings in config.toml file...
Start gaia on grpc port: 9092...
Balances for validator 'cosmos1ug9yrmkta2suar9cu4c3cfgh4d4gc4p5pekdne' @ 'tcp://localhost:26457'
balances: []
pagination:
  next_key: null
  total: "0"
Balances for user 'cosmos1d7llpvl2jt3ndwk3h3gjstxv7mlj9cjlhandyx' @ 'tcp://localhost:26457'
balances: []
pagination:
  next_key: null
  total: "0"
Balances for user 'cosmos1z7cvy9pz5emvshrqtplquyy2rw42fnn66m6fzt' @ 'tcp://localhost:26457'
balances: []
pagination:
  next_key: null
  total: "0"
ibc-0 initialized. Watch file /home/hussaini/Documents/ibc-rs/data/ibc-0.log to see its execution.
ibc-1 initialized. Watch file /home/hussaini/Documents/ibc-rs/data/ibc-1.log to see its execution.
ibc-2 initialized. Watch file /home/hussaini/Documents/ibc-rs/data/ibc-2.log to see its execution.
Building the Rust relayer...

Importing keys...
Success: Added key 'testkey' (cosmos1x9ueqshn8grl7qtj8agp6j359f4snm8yfgznd9) on chain ibc-0
Success: Added key 'user2' (cosmos1qhv4lcqm772h4d9k2dfrsx2gvyfeh0gr8jf4zv) on chain ibc-0
Success: Added key 'testkey' (cosmos1ph9cnw62mlgd997whfjdrgcuaq3a7usypt090f) on chain ibc-1
Success: Added key 'user2' (cosmos1t3mmfnrprxl46gn4cru654gktzu5t7uf64flst) on chain ibc-1
Success: Added key 'testkey' (cosmos1d7llpvl2jt3ndwk3h3gjstxv7mlj9cjlhandyx) on chain ibc-2
Success: Added key 'user2' (cosmos1z7cvy9pz5emvshrqtplquyy2rw42fnn66m6fzt) on chain ibc-2

$ hermes create client ibc-1 ibc-0
Success: CreateClient(
    CreateClient(
        Attributes {
            height: Height {
                revision: 1,
                height: 1858,
            },
            client_id: ClientId(
                "07-tendermint-0",
            ),
            client_type: Tendermint,
            consensus_height: Height {
                revision: 0,
                height: 1864,
            },
        },
    ),
)

$ hermes create client ibc-2 ibc-0
Jul 06 14:56:05.675  INFO ibc_relayer_cli::commands: using default configuration from '/home/hussaini/.hermes/config.toml'
Success: CreateClient(
    CreateClient(
        Attributes {
            height: Height {
                revision: 2,
                height: 1923,
            },
            client_id: ClientId(
                "07-tendermint-0",
            ),
            client_type: Tendermint,
            consensus_height: Height {
                revision: 0,
                height: 1942,
            },
        },
    ),
)

$ hermes tx raw upgrade-chain ibc-0 ibc-1 07-tendermint-0 10000000 300
Jul 06 15:00:01.034  INFO ibc_relayer_cli::commands: using default configuration from '/home/hussaini/.hermes/config.toml'
Jul 06 15:00:01.035  INFO ibc_relayer_cli::commands::tx::upgrade: Message UpdatePlanOptions { src_chain_config: ChainConfig { id: ChainId { id: "ibc-1", version: 1 }, rpc_addr: Url { inner: Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Ipv4(127.0.0.1)), port: Some(26557), path: "/", query: None, fragment: None }, scheme: Http, host: "127.0.0.1", port: 26557 }, websocket_addr: Url { inner: Url { scheme: "ws", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("localhost")), port: Some(26557), path: "/websocket", query: None, fragment: None }, scheme: WebSocket, host: "localhost", port: 26557 }, grpc_addr: Url { inner: Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Ipv4(127.0.0.1)), port: Some(9091), path: "/", query: None, fragment: None }, scheme: Http, host: "127.0.0.1", port: 9091 }, rpc_timeout: 10s, account_prefix: "cosmos", key_name: "testkey", store_prefix: "ibc", gas: Some(200000), fee_denom: "stake", fee_amount: Some(10), max_msg_num: None, max_tx_size: None, clock_drift: 5s, trusting_period: 1209600s, trust_threshold: TrustThresholdFraction { numerator: 1, denominator: 3 } }, dst_chain_config: ChainConfig { id: ChainId { id: "ibc-0", version: 0 }, rpc_addr: Url { inner: Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Ipv4(127.0.0.1)), port: Some(26657), path: "/", query: None, fragment: None }, scheme: Http, host: "127.0.0.1", port: 26657 }, websocket_addr: Url { inner: Url { scheme: "ws", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("localhost")), port: Some(26657), path: "/websocket", query: None, fragment: None }, scheme: WebSocket, host: "localhost", port: 26657 }, grpc_addr: Url { inner: Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Ipv4(127.0.0.1)), port: Some(9090), path: "/", query: None, fragment: None }, scheme: Http, host: "127.0.0.1", port: 9090 }, rpc_timeout: 10s, account_prefix: "cosmos", key_name: "testkey", store_prefix: "ibc", gas: Some(200000), fee_denom: "stake", fee_amount: Some(10), max_msg_num: None, max_tx_size: None, clock_drift: 5s, trusting_period: 1209600s, trust_threshold: TrustThresholdFraction { numerator: 1, denominator: 3 } }, src_client_id: ClientId("07-tendermint-0"), amount: 10000000, height_offset: 300 }
Success: []

$ gaiad tx gov vote 1 yes --home data/ibc-0/data/ --keyring-backend test --keyring-dir data/ibc-0/ --chain-id ibc-0 --from validator
{"body":{"messages":[{"@type":"/cosmos.gov.v1beta1.MsgVote","proposal_id":"1","voter":"cosmos1hp2dcwqycyfcpn9xtg4tn9hv65klfks33szfgv","option":"VOTE_OPTION_YES"}],"memo":"","timeout_height":"0","extension_options":[],"non_critical_extension_options":[]},"auth_info":{"signer_infos":[],"fee":{"amount":[],"gas_limit":"200000","payer":"","granter":""}},"signatures":[]}

confirm transaction before signing and broadcasting [y/N]: y
{"height":"2294","txhash":"A2AD88BA7C21B06773C9F2936EF506DF2670DE0309A7E3FBF00C5CD8AA8243F7","codespace":"","code":0,"data":"0A060A04766F7465","raw_log":"[{\"events\":[{\"type\":\"message\",\"attributes\":[{\"key\":\"action\",\"value\":\"vote\"},{\"key\":\"module\",\"value\":\"governance\"},{\"key\":\"sender\",\"value\":\"cosmos1hp2dcwqycyfcpn9xtg4tn9hv65klfks33szfgv\"}]},{\"type\":\"proposal_vote\",\"attributes\":[{\"key\":\"option\",\"value\":\"VOTE_OPTION_YES\"},{\"key\":\"proposal_id\",\"value\":\"3\"}]}]}]","logs":[{"msg_index":0,"log":"","events":[{"type":"message","attributes":[{"key":"action","value":"vote"},{"key":"module","value":"governance"},{"key":"sender","value":"cosmos1hp2dcwqycyfcpn9xtg4tn9hv65klfks33szfgv"}]},{"type":"proposal_vote","attributes":[{"key":"option","value":"VOTE_OPTION_YES"},{"key":"proposal_id","value":"1"}]}]}],"info":"","gas_wanted":"200000","gas_used":"43716","tx":null,"timestamp":""}

$ gaiad query gov proposal 1 --home data/ibc-0/
content:
  '@type': /cosmos.upgrade.v1beta1.SoftwareUpgradeProposal
  description: upgrade the chain software and unbonding period
  plan:
    height: "2564"
    info: upgrade the chain software and unbonding period
    name: test
    time: "0001-01-01T00:00:00Z"
    upgraded_client_state:
      '@type': /ibc.lightclients.tendermint.v1.ClientState
      allow_update_after_expiry: false
      allow_update_after_misbehaviour: false
      chain_id: ibc-0
      frozen_height:
        revision_height: "0"
        revision_number: "0"
      latest_height:
        revision_height: "2565"
        revision_number: "0"
      max_clock_drift: 0s
      proof_specs:
      - inner_spec:
          child_order:
          - 0
          - 1
          child_size: 33
          empty_child: null
          hash: SHA256
          max_prefix_length: 12
          min_prefix_length: 4
        leaf_spec:
          hash: SHA256
          length: VAR_PROTO
          prefix: AA==
          prehash_key: NO_HASH
          prehash_value: SHA256
        max_depth: 0
        min_depth: 0
      - inner_spec:
          child_order:
          - 0
          - 1
          child_size: 32
          empty_child: null
          hash: SHA256
          max_prefix_length: 1
          min_prefix_length: 1
        leaf_spec:
          hash: SHA256
          length: VAR_PROTO
          prefix: AA==
          prehash_key: NO_HASH
          prehash_value: SHA256
        max_depth: 0
        min_depth: 0
      trust_level:
        denominator: "0"
        numerator: "0"
      trusting_period: 0s
      unbonding_period: 1440000s
      upgrade_path:
      - upgrade
      - upgradedIBCState
  title: upgrade_ibc_clients
deposit_end_time: "2021-07-06T09:38:52.973644456Z"
final_tally_result:
  abstain: "0"
  "no": "0"
  no_with_veto: "0"
  "yes": "100000000000"
proposal_id: "1"
status: PROPOSAL_STATUS_PASSED
submit_time: "2021-07-06T09:35:32.973644456Z"
total_deposit:
- amount: "10000000"
  denom: stake
voting_end_time: "2021-07-06T09:38:52.973644456Z"
voting_start_time: "2021-07-06T09:35:32.973644456Z"
```

But when I try the `upgrade clients` command I get the below error ->
```bash
$ hermes upgrade clients ibc-0
Jul 06 15:11:36.189  INFO ibc_relayer_cli::commands: using default configuration from '/home/hussaini/.hermes/config.toml'
Jul 06 15:11:36.270  INFO ibc_relayer::foreign_client: [ibc-1 -> ibc-0:07-tendermint-0] upgrade Height: 1-2465
Jul 06 15:11:36.323  INFO ibc_relayer::foreign_client: [ibc-2 -> ibc-0:07-tendermint-0] upgrade Height: 2-2454
Error: .
├─.
├─.
│ └─ query error: error raised while updating client: failed querying client state on dst chain 07-tendermint-0 with error: Query error occurred (failed to query for client state): error converting message type into domain type: the client state was not found
└─.
  └─ query error: error raised while updating client: failed querying client state on dst chain 07-tendermint-0 with error: Query error occurred (failed to query for client state): error converting message type into domain type: the client state was not found
```

Interestingly, when I change the `ForeignClient::restore()` with `ForeignClient::find()` based on what the `upgrade client` command is doing, the `upgrade clients` command works without a problem. Still trying to figure out why this is the case.

hu55a1n1 commented:
So, there was a bug in the code that @adizere identified during our call - the order of args for the `ForeignClient::restore()` function was wrong. I fixed that and now I have a different error -> 
```bash
$ hermes upgrade clients ibc-0
    Finished release [optimized] target(s) in 0.21s
     Running `target/release/hermes upgrade clients ibc-0`
Jul 08 17:39:02.511  INFO using default configuration from '/home/hussaini/.hermes/config.toml'
Jul 08 17:39:02.517 TRACE subscribing to query: tm.event = 'Tx' chain.id=ibc-0
Jul 08 17:39:02.517 TRACE subscribing to query: tm.event = 'NewBlock' chain.id=ibc-0
Jul 08 17:39:02.518 TRACE subscribed to all queries chain.id=ibc-0
>>> upgrading clients ChainId { id: "ibc-0", version: 0 } -> ChainId { id: "ibc-1", version: 1 }
Jul 08 17:39:02.518 DEBUG starting event monitor chain.id=ibc-0
Jul 08 17:39:02.522 TRACE subscribing to query: tm.event = 'Tx' chain.id=ibc-1
Jul 08 17:39:02.523 TRACE subscribing to query: tm.event = 'NewBlock' chain.id=ibc-1
Jul 08 17:39:02.523 TRACE subscribed to all queries chain.id=ibc-1
Jul 08 17:39:02.523 DEBUG starting event monitor chain.id=ibc-1
>>> upgrading ChainId { id: "ibc-0", version: 0 } -> ChainId { id: "ibc-1", version: 1 }: ClientId("07-tendermint-0")
Jul 08 17:39:02.526  INFO [ibc-0 -> ibc-1:07-tendermint-0] upgrade Height: 0-225
Jul 08 17:39:02.528 TRACE light client verification trusted=0-73 target=0-225
Jul 08 17:39:02.535 TRACE adjusting headers with 0 supporting headers trusted=0-73 target=225
Jul 08 17:39:02.535 TRACE fetching header height=0-74
Jul 08 17:39:02.539 DEBUG [ibc-0 -> ibc-1:07-tendermint-0] MsgUpdateAnyClient from trusted height 0-73 to target height 0-225
Jul 08 17:39:02.541 DEBUG [ibc-0 -> ibc-1:07-tendermint-0] upgraded client state Tendermint(ClientState { chain_id: ChainId { id: "ibc-0", version: 0 }, trust_level: TrustThresholdFraction { numerator: 0, denominator: 0 }, trusting_period: 0ns, unbonding_period: 1440000s, max_clock_drift: 0ns, frozen_height: Height { revision: 0, height: 0 }, latest_height: Height { revision: 0, height: 383 }, upgrade_path: ["upgrade", "upgradedIBCState"], allow_update: AllowUpdate { after_expiry: false, after_misbehaviour: false } })
>>> upgrading clients ChainId { id: "ibc-0", version: 0 } -> ChainId { id: "ibc-2", version: 2 }
Jul 08 17:39:02.559 TRACE subscribing to query: tm.event = 'Tx' chain.id=ibc-2
Jul 08 17:39:02.559 TRACE subscribing to query: tm.event = 'NewBlock' chain.id=ibc-2
Jul 08 17:39:02.559 TRACE subscribed to all queries chain.id=ibc-2
Jul 08 17:39:02.560 DEBUG starting event monitor chain.id=ibc-2
>>> upgrading ChainId { id: "ibc-0", version: 0 } -> ChainId { id: "ibc-2", version: 2 }: ClientId("07-tendermint-0")
Jul 08 17:39:02.562  INFO [ibc-0 -> ibc-2:07-tendermint-0] upgrade Height: 0-225
Jul 08 17:39:02.566 TRACE light client verification trusted=0-77 target=0-225
Jul 08 17:39:02.571 TRACE adjusting headers with 0 supporting headers trusted=0-77 target=225
Jul 08 17:39:02.571 TRACE fetching header height=0-78
Jul 08 17:39:02.574 DEBUG [ibc-0 -> ibc-2:07-tendermint-0] MsgUpdateAnyClient from trusted height 0-77 to target height 0-225
Jul 08 17:39:02.577 DEBUG [ibc-0 -> ibc-2:07-tendermint-0] upgraded client state Tendermint(ClientState { chain_id: ChainId { id: "ibc-0", version: 0 }, trust_level: TrustThresholdFraction { numerator: 0, denominator: 0 }, trusting_period: 0ns, unbonding_period: 1440000s, max_clock_drift: 0ns, frozen_height: Height { revision: 0, height: 0 }, latest_height: Height { revision: 0, height: 383 }, upgrade_path: ["upgrade", "upgradedIBCState"], allow_update: AllowUpdate { after_expiry: false, after_misbehaviour: false } })
Error: .
├─.
│ └─ query error: failed while trying to upgrade client id 07-tendermint-0 with error: failed while fetching from chain ibc-0 the upgraded client consensus state: GRPC error: status: InvalidArgument, message: "upgraded consensus state not found in store for height: 225: invalid request: invalid request", details: [], metadata: MetadataMap { headers: {"content-type": "application/grpc"} }
└─.
  └─ query error: failed while trying to upgrade client id 07-tendermint-0 with error: failed while fetching from chain ibc-0 the upgraded client consensus state: GRPC error: status: InvalidArgument, message: "upgraded consensus state not found in store for height: 225: invalid request: invalid request", details: [], metadata: MetadataMap { headers: {"content-type": "application/grpc"} }
```
After much debugging, I realised that this error is also happening on `master` with `hermes upgrade client ibc-1 07-tendermint-0`, this means that the [Testing Client Upgrade](https://hermes.informal.systems/commands/upgrade/test.html) procedure in the hermes guide is not working on latest `master`. Strangely, the problem resolves by itself after some time.

adizere reviewed (approved):
One small comment I have is that there is an error message that could be improved, namely:

> Error: .
├─.
│ └─ query error: failed while trying to upgrade client id 07-tendermint-1 with error: failed while fetching from chain ibc-1 the upgraded client state: Empty response value
└─.
  └─ query error: failed while trying to upgrade client id 07-tendermint-3 with error: failed while fetching from chain ibc-1 the upgraded client state: Empty response value

To reproduce this error, simply invoke `hermes upgrade clients ibc-1` so that the chain `ibc-1` did __not undergo an upgrade__. The error simply states that chain `ibc-1` defines no upgrade client state, but the user has no idea where is client `id 07-tendermint-1` hosted -- on which chain? In `ForeignClient` terminology, this would be the destination chain.

To be more specific, all that's needed is to change the error type `ClientUpgrade` similartly to this:

```diff
diff --git a/relayer/src/foreign_client.rs b/relayer/src/foreign_client.rs
index 3b63e003..5e77d647 100644
--- a/relayer/src/foreign_client.rs
+++ b/relayer/src/foreign_client.rs
@@ -61,8 +61,8 @@ pub enum ForeignClientError {
     #[error("cannot run misbehaviour: {0}")]
     MisbehaviourExit(String),

-    #[error("failed while trying to upgrade client id {0} with error: {1}")]
-    ClientUpgrade(ClientId, String),
+    #[error("failed while trying to upgrade client id {0} hosted on chain id {1} with error: {2}")]
+    ClientUpgrade(ClientId, ChainId, String),
```

Note that after modifying `ClientUpgrade` definition you may need to adapt a few parts of the code in `ForeignClient`.

This is just a nice to have, so feel free to open a follow-up issue to capture this problem and merge this PR as it is! Or alternatively fix it in this PR, as you wish.

Thanks Shoaib, cool work!

hu55a1n1 commented:
> This is just a nice to have, so feel free to open a follow-up issue to capture this problem and merge this PR as it is! Or alternatively fix it in this PR, as you wish.

Thanks for the suggestions and the detailed comment @adizere! I agree that having the host chain id in the error will be very helpful. 👌 I have applied your suggestions to this PR itself.

adizere commented on relayer/src/foreign_client.rs:
```diff
@@ -211,9 +213,12 @@ impl ForeignClient {
-            .map_err(|e| ForeignClientError::ClientUpgrade(self.id.clone(), format!(
-                "failed while fetching from chain {} the upgraded client consensus state: {}", self.src_chain.id(), e)))
-            ?;
+            .map_err(|e| ForeignClientError::ClientUpgrade(self.id.clone(),
```
Beautiful work, Shoaib!
